### PR TITLE
Add support for text pair classification (ASNQ) and ranking (MSMarco)

### PR DESCRIPTION
**Pull request by brandenchan**
This pull request adds support text pair classification where the input is two separate sentences and the output is some form of classification. In the case of ASNQ, the prediction states whether a question can be answered by a certain passage.

### CODE REVIEW

**Comment by brandenchan** [comment]
Column and feature naming is still a bit messy (tokens and tokens_b). Also still need to assess whether the TextPairClassificationProcessor should be a stand alone class or be incorporated into TextClassificationProcessor.

**Comment by brandenchan** [comment]
There are examples now for MSMarco and ASNQ (binary). There is an experiment config, only for ASNQ since MSMarco requires a lot of preprocessing and separate metric scripts.

File: examples/passage_ranking.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    device, n_gpu = initialize_device_settings(use_cuda=True)
+    n_epochs = 2
+    batch_size = 64
+    evaluate_every = 500
+    lang_model = "bert-base-cased"
+    label_list = ["0", "1"]
+    train_filename = "train.tsv"
+    dev_filename = "dev.tsv"
+    generate_data = True
+    data_dir = Path("../data/msmarco_passage")
```
Comment (tholor) [inline examples/passage_ranking.py]: Maybe add a link here to the dataset that users should download

File: examples/passage_ranking.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    predictions_filename = "predictions.txt"
+    train_source_filename = "triples.train.small.tsv"
+    qrels_filename = "qrels.dev.tsv"
+    queries_filename = "queries.dev.tsv"
+    passages_filename = "collection.tsv"
+    top1000_filename = "top1000.dev"
+
+
+    # 0. Preprocess and save MSMarco data in a format that can be ingested by FARM models. Only needs to be done once!
+    if generate_data:
```
Comment (tholor) [inline examples/passage_ranking.py]: Would it make sense to give here a quick example of the target format for the TextPairClassificationProcessor? (e.g text_a      text_b     label). Might help users to use it for their own dataset (which is not msmarco)

File: farm/data_handler/input_features.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def sample_to_features_text(
         tokens_b,
         add_special_tokens=True,
         max_length=max_seq_len,
-        truncation_strategy='do_not_truncate' # We've already truncated our tokens before
+        truncation_strategy='only_second'
```
Comment (tholor) [inline farm/data_handler/input_features.py]: Why do we need this change here?

File: farm/data_handler/processor.py [hunk 4]
```diff
@@ -491,6 +493,32 @@ def _sample_to_features(self, sample) -> dict:
         )
         return features
 
+class TextPairClassificationProcessor(TextClassificationProcessor):
```
Comment (tholor) [inline farm/data_handler/processor.py]: Please add a short docstring

File: farm/data_handler/processor.py [hunk 5]
```diff
@@ -1069,6 +1099,7 @@ def __init__(
         self.delimiter = delimiter
         self.quote_char = quote_char
         self.skiprows = skiprows
+        self.label_remapping = label_remapping
```
Comment (tholor) [inline farm/data_handler/processor.py]: What is this one doing here? If we need it, we should add a docstring

File: farm/data_handler/processor.py (hunk 4 above)
Comment (tholor) [reply farm/data_handler/processor.py]: It should be clear what input format is expected (e.g. colum names = "text"; "text_b", "label")

File: farm/evaluation/msmarco_passage_farm.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+from farm.evaluation.msmarco_passage_official import compute_metrics_from_files
+import os
+import pandas as pd
+
+
+def msmarco_evaluation(preds_file, dev_file, qrels_file, output_file):
+    """
```
Comment (tholor) [inline farm/evaluation/msmarco_passage_farm.py]: One line describing what type of evaluation we are doing here (e.g. with link to task website)

Comment (tholor) [reply farm/evaluation/msmarco_passage_farm.py]: Also make the relation to the official script clearer. As I understood it, this is a wrapper to bring the predictions into the right shape and then run the official eval function "compute_metrics_from_files"?

File: farm/data_handler/processor.py [hunk 7]
```diff
@@ -491,6 +493,32 @@ def _sample_to_features(self, sample) -> dict:
...
+            rename_columns=column_mapping,
+            filename=file,
+            delimiter=self.delimiter,
+            skiprows=self.skiprows,
+            proxies=self.proxies,
+        )
+        return dicts
+
+    def _dict_to_samples(self, dictionary: dict, **kwargs) -> [Sample]:
+        tokenized_question = tokenize_with_metadata(dictionary["text"], self.tokenizer)
```
Comment (tholor) [inline farm/data_handler/processor.py]: I would keep the naming here generic, e.g. "tokenized_a" and tokenized_b

**Pull request merged**